### PR TITLE
Split TypeName into Internal and External variants

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Part of an ongoing refactoring to make our object model cleaner, making illegal state unrepresentable.

Introduces `InternalTypeName` and `ExternalTypeName` as the two possible implementations of `TypeName`. 

`ExternalTypeName` may only reference an `ExternalPackageReference`.

For now `InternalTypeName` still uses `PackageReference` because that change has a wider impact and I didn't want to bloat this PR further.

Also renamed `LocalLikePackageReference` to `InternalPackageReference`, as it's a useful common denominator across `LocalPackageReference`, `StoragePackageReference` and `SubPackageReference`.

**Prerequisites**

- [x] PR #3174 - Covert TypeName to an interface
- [x] PR #3175 - Add support for sub packages

**Special notes for your reviewer**:

There are a few spots where we now have some awkward casts; these are temporary will be fixed up in future PRs.
We'll also end up simplifying the `TypeName` and `PackageReference` interfaces.

Requires merge of the prerequisites before CI will run, due to branch targeting.

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: v2/tools/generator/internal/astmodel/internal_type_name.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+			codeGenerationContext.currentPackage))
+	}
+
+	return astbuilder.Selector(dst.NewIdent(packageName), tn.Name())
+}
+
+// AsZero renders an expression for the "zero" value of the type.
+// The exact thing we need to generate depends on the actual type we reference
+func (tn InternalTypeName) AsZero(definitions TypeDefinitionSet, ctx *CodeGenerationContext) dst.Expr {
+	if IsExternalPackageReference(tn.packageReference) {
```
Comment (matthchr) [inline v2/tools/generator/internal/astmodel/internal_type_name.go:83]: Is this still possible?

File: v2/tools/generator/internal/astmodel/internal_type_name.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+		// (we might not actually use this, if the property is optional, but we still need to generate the right thing)
+
+		packageName := ctx.MustGetImportedPackageName(tn.packageReference)
+		return &dst.SelectorExpr{
+			X:   dst.NewIdent(packageName),
+			Sel: dst.NewIdent(fmt.Sprintf("%s{}", tn.Name())),
+		}
+	}
+
+	actualType, err := definitions.FullyResolve(tn)
```
Comment (matthchr) [inline v2/tools/generator/internal/astmodel/internal_type_name.go:94]: Do we want some shared code for bits of this implementation and the `external_type_name`?

Comment (matthchr) [reply v2/tools/generator/internal/astmodel/internal_type_name.go:94]: This looks quite similar to what's over there.

File: v2/tools/generator/internal/astmodel/internal_package_reference.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+package astmodel
+
+// InternalPackageReference describes a package reference that points to a local package (either a storage package
+// or a standard one). It can be used to abstract across the exact package type (storage vs local)
+type InternalPackageReference interface {
+	PackageReference
+
+	// LocalPathPrefix returns the prefix (everything up to the group name)
+	LocalPathPrefix() string
+	Group() string
```
Comment (matthchr) [inline v2/tools/generator/internal/astmodel/internal_package_reference.go:15]: You embed PackageReference but then redefine a bunch of its methods here too? Not following why?

File: v2/tools/generator/internal/codegen/pipeline/add_operator_spec.go [hunk 4]
```diff
@@ -326,7 +326,7 @@ func newOperatorSpecBuilder(
 
 func (b *operatorSpecBuilder) newEmptyOperatorSpec() astmodel.TypeDefinition {
 	name := b.idFactory.CreateIdentifier(b.resource.Name().Name()+"OperatorSpec", astmodel.Exported)
-	operatorSpecTypeName := b.resource.Name().WithName(name)
+	operatorSpecTypeName := b.resource.Name().WithName(name).(astmodel.InternalTypeName)
```
Comment (matthchr) [inline v2/tools/generator/internal/codegen/pipeline/add_operator_spec.go:329]: To be improved in a future PR, I assume?

File: v2/tools/generator/internal/codegen/pipeline/load_types.go [hunk 5]
```diff
@@ -94,7 +94,7 @@ func LoadTypes(
 
 				scope := categorizeResourceScope(resourceInfo.ARMURI)
 				resourceType = resourceType.WithScope(scope)
-				resourceDefinition := astmodel.MakeTypeDefinition(resourceName, resourceType)
+				resourceDefinition := astmodel.MakeTypeDefinition(resourceName.(astmodel.InternalTypeName), resourceType)
```
Comment (matthchr) [inline v2/tools/generator/internal/codegen/pipeline/load_types.go:97]: Will improve later? icky atm.

**Review by matthchr — APPROVED** [review]

File: v2/tools/generator/internal/astmodel/internal_package_reference.go (hunk 3 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/astmodel/internal_package_reference.go:15]: This used to be `LocalLikePackageReference` - I'm pulling it into the object model fully. I'll simplify the declaration.

File: v2/tools/generator/internal/codegen/pipeline/load_types.go (hunk 5 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/codegen/pipeline/load_types.go:97]: Yup. There's a lot of this that will go away.

**Pull request merged**